pkg/api/steve/supportconfigs/generator.go

```diff
@@ -0,0 +1,168 @@
+
+	v1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/sirupsen/logrus"
+	authzv1 "k8s.io/api/authorization/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	authv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+)
+
+const (
+	// Endpoint The endpoint that this URL is accessible at - used for routing and for SARs
+	Endpoint            = "/v1/generateSUSERancherSupportConfig"
+	cspAdapterNamespace = "cattle-csp-system"
+	cspAdapterConfigmap = "csp-config"
+	tarContentType      = "application/x-tar"
+	logPrefix           = "support-config-generator"
+)
+
+type GeneratorHandler struct {
+	ConfigMaps           v1.ConfigMapInterface
+	SubjectAccessReviews authv1.SubjectAccessReviewInterface
+	ConfigHandler        supportConfigHandler
+}
+
+func NewGeneratorHandler(scaledContext *config.ScaledContext) GeneratorHandler {
+	return GeneratorHandler{
+		ConfigMaps:           scaledContext.Core.ConfigMaps(cspAdapterNamespace),
+		SubjectAccessReviews: scaledContext.K8sClient.AuthorizationV1().SubjectAccessReviews(),
+		ConfigHandler:        supportConfigHandler{},
+	}
+}
+
+func (h *GeneratorHandler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	authorized, err := h.authorize(request)
+	if err != nil {
+		_, _ = writer.Write([]byte("forbidden"))
+		writer.WriteHeader(http.StatusForbidden)
+		logrus.Errorf("[%s] Failed to authorize user with error: %s", logPrefix, err.Error())
+		return
+	}
+	if !authorized {
+		_, _ = writer.Write([]byte("forbidden"))
+		writer.WriteHeader(http.StatusForbidden)
+		return
+	}
+	logrus.Infof("[%s] Generating supportconfig", logPrefix)
+	archive, err := h.generateSupportConfig()
+	logrus.Infof("[%s] Done Generating supportconfig", logPrefix)
+	if err != nil {
+		logrus.Errorf("[%s] Error when generating support config %v", logPrefix, err)
+		_, _ = writer.Write([]byte("Internal Server Error"))
+		writer.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	writer.Header().Set("Content-Type", tarContentType)
+	writer.Header().Set("Content-Disposition", "attachment; filename=\"supportConfig.tar\"")
+	n, err := io.Copy(writer, archive)
+	logrus.Debugf("[%s] wrote %v bytes in archive response", logPrefix, n)
+}
+
+// authorize checks to see if the user can get the non-resource url of Endpoint. If not, returns an error, otherwise nil
+func (h *GeneratorHandler) authorize(r *http.Request) (bool, error) {
+	userInfo, ok := request.UserFrom(r.Context())
+	if !ok {
+		return false, fmt.Errorf("unable to extract user info from context")
+	}
+	response, err := h.SubjectAccessReviews.Create(r.Context(), &authzv1.SubjectAccessReview{
+		Spec: authzv1.SubjectAccessReviewSpec{
+			// This auth method prevents collision with future k8s resources and allows this to be added to custom rules
+			NonResourceAttributes: &authzv1.NonResourceAttributes{
```

Comment:
> Now that I understand that the entire supportconfig will be contained in one configmap with a predictable name and namespace
One correction here - rancher still produces some values (i.e. arch/product version) on it's own. The adapter does a lot of the 

This would make a bit more sense, but there's a couple of edge cases that make me unsure of this route. The big issue here is that the configmap/csp adapter namespace is only present when the adapter is installed, but we still need to be able to produce a limited supportconfig even without it. I'm unsure about the ability of a user to grant access to a configmap in a namespace which doesn't exist.

The other issue would be controlling access through custom role. Given that our global roles map to cluster roles in the local cluster, I think that it would be challenging to grant access to a custom admin role. You might be able to get away with it by granting access to every configmap with that hardcoded name in every namespace, but that will be challenging for the customer to figure out and would grant more access than intended. 

That being said, I'm flexible on this since this does meet the requirements and is less abusive of the k8s api. Let me know what your thoughts are.

Reply:
One thing that is messing with our design is the idea that any Rancher in the future (non-CSP included) should have the "get support config" button which downloads a support config tar with logs included. So that case makes more sense for SAR route, unless we pre-package this configmap.